Vesper-Agents-Suite/xy-safe-borrow-agent/src/agent.ts

```diff
@@ -0,0 +1,166 @@
+const tracker: TimeTracker = new TimeTracker();
+const vesperFetcher: VesperFetcher = new VesperFetcher(web3);
+let cacheTime: number = 3600000; // one hour in milliseconds
+const DECIMAL18 = new BigNumber('1e18')
+const MAX_BPS = 10000
+
+export const setCacheTime = (time: number) => {
+  cacheTime = time;
+}
+
+const fetchXYStrategies = async (
+  blockNumber: string
+): Promise<string[]> => {
+  if (blockNumber != "latest" && cache.get(blockNumber) !== undefined)
+    return cache.get(blockNumber) as string[];
+  let [V3] = await Promise.all([
+    vesperFetcher.getStrategiesV3(blockNumber)
+  ]);
+  V3 = Array.from(new Set<string>(V3));
+  const valueV3promise = V3.map(async (strategy: string) => {
+    const strategyContract = new web3.eth.Contract(
+      STRATEGY_ABI,
+      strategy,
+    );
+
+    const name: string = await strategyContract.methods.NAME().call({}, blockNumber);
+    // Compound or it's forked XY Strategies 
+    if (!name.includes(XY_STRATEGY_NAME) || name.includes(Aave)) return BigInt(0);
+
+    // get borrow token
+    const borrowCTokenAddress: string = await strategyContract.methods.borrowCToken().call({}, blockNumber);
+    const borrowCToken = new web3.eth.Contract(
+      CTOKEN_ABI,
+      borrowCTokenAddress,
+    );
+
+    // get supply token
+    const supplyCTokenAddress: string = await strategyContract.methods.supplyCToken().call({}, blockNumber);
+    const supplyCToken = new web3.eth.Contract(
+      CTOKEN_ABI,
+      supplyCTokenAddress,
+    );
+
+    // get comptroller
+    const comptrollerAddress: string = await strategyContract.methods.comptroller().call({}, blockNumber);
+    const actualComptrollerAbi = name.includes(TRADER_JOE) ? TRADER_JOE_ABI : COMPOUND_ABI
+    const comptrollerContract = new web3.eth.Contract(
+      actualComptrollerAbi,
+      comptrollerAddress,
+    );
+
+    // get Oracle
+    const oracleAddress: string = await comptrollerContract.methods.oracle().call({}, blockNumber);
+    const oracleContract = new web3.eth.Contract(
+      ORACLE_ABI,
+      oracleAddress,
+    );
+
+    // get prices
+    const borrowTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(borrowCTokenAddress).call({}, blockNumber));
+    const supplyTokenPrice = new BigNumber(await oracleContract.methods.getUnderlyingPrice(supplyCTokenAddress).call({}, blockNumber));
+    const borrowed = new BigNumber(await borrowCToken.methods.borrowBalanceStored(strategy).call({}, blockNumber));
+    const supplyBalance = new BigNumber(await supplyCToken.methods.balanceOf(strategy).call({}, blockNumber));
+    const supplyTokenExchangeRate = new BigNumber(await supplyCToken.methods.exchangeRateStored().call({}, blockNumber));
+    const cf = (await comptrollerContract.methods.markets(supplyCTokenAddress).call({}, blockNumber)).collateralFactorMantissa
+    const maxBorrowLimit = new BigNumber(await strategyContract.methods.maxBorrowLimit().call({}, blockNumber));
```

Comment:
This is not good practice that you are using 10 await statement. Most of these can be fired simultaneously by using Promise.all()